### PR TITLE
mkosi: take a BSD lock on the workspace directory while running

### PR DESCRIPTION
**Pull request by poettering**
This makes use of the new logic here:

https://github.com/systemd/systemd/pull/11482

(And on older systems has no effect, but that shouldn't hurt either)

Fixes: #252

### CODE REVIEW

File: mkosi [hunk 1]
```diff
@@ -4307,6 +4312,7 @@ def build_stuff(args: CommandLineArguments) -> None:
     if root_hash is not None:
         print_step(f'Root hash is {root_hash}.')
 
+    del dir_fd
```
Comment (lucaswerkmeister) [inline mkosi]: Shouldn’t we `os.close(fd)` instead? As far as I can tell, `os.open` returns a plain `int`, so deleting it won’t have any special effect.

File: mkosi [hunk 2]
```diff
@@ -4243,6 +4243,11 @@ def build_stuff(args: CommandLineArguments) -> None:
     setup_package_cache(args)
     workspace = setup_workspace(args)
 
+    # Make sure tmpfiles' aging doesn't interfere with our workspace
+    # while we are working on it.
+    dir_fd = os.open(workspace.name, os.O_RDONLY|os.O_DIRECTORY|os.O_CLOEXEC)
+    fcntl.flock(dir_fd, fcntl.LOCK_EX)
```
Comment (lucaswerkmeister) [inline mkosi]: Should this include `fcntl.LOCK_NB`? If a lock is already held on the directory for some reason, it’s probably better for us to proceed (and risk that the other lock is released before we’re done, making us vulnerable to the tmpfiles cleanup again) than block until the other lock is released (we have no idea when that would be).

Comment (poettering) [reply mkosi]: well, the idea is that if tmpfiles is currently iterating through our directory, then we should wait until it is finished with that before we start using the directory. hence it should be a blocking open

File: mkosi (hunk 1 above)
Comment (poettering) [reply mkosi]: indeed

**Comment by poettering** [comment]
force pushed a new version. only change: use "os.close()" instead of "del" for the fd.

PTAL!

**Pull request merged**